src/data-workspace/category-combo-table/category-combo-table-header.js

```diff
@@ -0,0 +1,130 @@
+import i18n from '@dhis2/d2-i18n'
+import { TableRowHead, TableCellHead } from '@dhis2/ui'
+import cx from 'classnames'
+import PropTypes from 'prop-types'
+import React from 'react'
+import { useMetadata, selectors } from '../../metadata/index.js'
+import { useActiveCell } from '../data-entry-cell/index.js'
+import styles from './category-combo-table.module.css'
+import { PaddingCell } from './padding-cell.js'
+import { TotalHeader } from './total-cells.js'
+
+export const CategoryComboTableHeader = ({
+    dataElements,
+    renderRowTotals,
+    paddingCells,
+    categoryOptionCombos,
+    categories,
+}) => {
+    const { data: metadata } = useMetadata()
+    const { deId: activeDeId, cocId: activeCocId } = useActiveCell()
+
+    // Computes the span and repeats for each columns in a category-row.
+    // Repeats are the number of times the options in a category needs to be rendered per category-row
+    let catColSpan = categoryOptionCombos.length
+    const rowToColumnsMap = categories.map((c) => {
+        const categoryOptions = selectors.getCategoryOptionsByCategoryId(
+            metadata,
+            c.id
+        )
+        const nrOfOptions = c.categoryOptions.length
+        // catColSpan should always be equal to nrOfOptions in last iteration
+        // unless anomaly with categoryOptionCombo-generation server-side
+        if (nrOfOptions > 0 && catColSpan >= nrOfOptions) {
+            // calculate colSpan for current options
+            // this is the span for each option, not the "total" span of the row
+            catColSpan = catColSpan / nrOfOptions
+            // when table have multiple categories, options need to be repeated for each disaggregation "above" current-category
+            const repeat =
+                categoryOptionCombos.length / (catColSpan * nrOfOptions)
+
+            const columnsToRender = new Array(repeat)
+                .fill(0)
+                .flatMap(() => categoryOptions)
+
+            return {
+                span: catColSpan,
+                columns: columnsToRender,
+                category: c,
+            }
+        } else {
+            console.warn(
+                `Category ${c.displayFormName} malformed. Number of options: ${nrOfOptions}, span: ${catColSpan}`
+            )
+        }
+        return c
+    })
+
+    // Is the active cell in this cat-combo table? Check to see if active
+    // data element is in this CCT
+    const isThisTableActive = dataElements.some(({ id }) => id === activeDeId)
+
+    // Find if this column header includes the active cell's column
+    const isHeaderActive = (headerIdx, headerColSpan) => {
+        const activeCellColIdx = categoryOptionCombos.findIndex(
+            (coc) => activeCocId === coc.id
+        )
+        const idxDiff = activeCellColIdx - headerIdx * headerColSpan
+        return isThisTableActive && idxDiff < headerColSpan && idxDiff >= 0
+    }
```

Comment:
I was just wondering why we also need to `isThisTableActive` const. Is it possible that a form contains several tables with the same dataElements?

Look at this very naively, I would expect that every cell will be unique so we don't need this variable per se. If you have added that `isThisTableActive` variable to make the code more performant, then you should add it as an early return in the `isHeaderActive` function.

Having said all of that, if you agree with my earlier suggestion, about moving the active/inactive logic down to the actual header cell, then this code is going to change anyway.

Reply:
> I did some refactoring to not need to pass all `dataElements` to this component. So I pass a function `checkTableActive` instead. But yes, it's needed because section-forms render multiple tables, and we only want to highlight the active table.
> 
> > Is it possible that a form contains several tables with the same dataElements?
> 
> No, and that's why this fact is used to check if the current table is "active" or not.

Partly from your explanation above, and by looking at the code a bit more closely, I get this now 👍 .

---

> > then you should add it as an early return in the isHeaderActive function.
> 
> I don't really understand what you mean here?

I was mainly just figuring out what was happening, and my suggestion about returning early was basically just this (see the comment):

```js
    const isHeaderActive = (headerIdx, headerColSpan) => {
        if (!checkTableActive(activeDeId)) {
            return false
        }
        // by returning early you can avoid this iteration
        const activeCellColIdx = categoryOptionCombos.findIndex(
            (coc) => activeCocId === coc.id
        )
        const idxDiff = activeCellColIdx - headerIdx * headerColSpan
        return (idxDiff < headerColSpan && idxDiff >= 0)
    }
```

But this is very much a micro-optimisation, so I'm very happy to skip this.